Test expiration

jmsanders opened the pull request:
I don't think expiration works as intended (or at least as I would expect it).

The response keys are correctly expired, but nothing is ever removed from the idempotency keys set. This has two consequences:

- once a response key expires, requests using the same idempotency key will return 409 forever
- the idempotency keys set grows without bound

Instead, I would expect expiration to also apply to the idempotency keys set. Once the expiration window has passed, a new request with the same idempotency key should issue a new request.

This commit only introduces the tests without implementing a solution. For MemoryBackend, I'd propose storing the expiry alongside the idempotency key. For RedisBackend, I'd propose something along the lines of https://github.com/redis/redis/issues/135#issuecomment-2361996 using a sorted set to purge expired idempotency keys. But I wanted to run the change by you before taking action on either approach.

jmsanders committed:
Test expiration

I don't think expiration works as intended (or at least as I would
expect it to).

The response keys are correctly expired, but nothing is ever removed
from the idempotency keys set. This has two consequences:

- once a response key expires, requests using the same idempotency key
  will return 409 forever
- the idempotency keys set grows without bound

Instead, I would expect expiration to also apply to the idempotency keys
set. Once the expiration window has passed, a new request with the same
idempotency key should issue a new request.

This commit only introduces the tests without implementing a solution.
For MemoryBackend, I'd propose storing the expiry alongside the
idempotency key. For RedisBackend, I'd propose something along the lines
of https://github.com/redis/redis/issues/135#issuecomment-2361996 using
a sorted set to purge expired idempotency keys. But I wanted to run the
change by you before taking action on either approach.

sondrelg commented:
Yeah I agree this seems like an implementation we've forgotten to account for. I like the look of the redis suggestion, so feel free to finish the implementation along those lines 👍 Just ping me when you consider this ready and I'll take another look 🙂

jmsanders commented:
Great - I'll throw something up later this week.

jmsanders committed:
Revert "Test expiration"

This reverts commit 9057f07394f0a61cf77c4c4dd7e9fd9dabc503f6.

jmsanders committed:
Expire idempotency keys

Previously, idempotency keys were never being expired. This had two
consequences:

- once a response key expires, requests using the same idempotency
  key will return 409 forever
- the idempotency keys set grows without bound

There's a new `expire_idempotency_keys` abstract method that
implementations of Backend are responsible for implementing. The
middleware calls it on every request.

For the Memory backend, we now store idempotency keys in a dict intead
of a set. The value of each key is its expiration time.

We then iterate over all of the items in the dict and
remove the ones with an expiration earlier than the current time. This
operation isn't very efficient because it iterates over the entire dict,
but given that the Memory backend is only meant to be used in local
testing, I don't think it's worth optimizing further.

For the RedisBackend, we now store idempotency keys in a sortedset
instead of a set. The "score" of each key is its expiration time. This
matches the Redis project's official recommendation on how to expire
items in a set: https://github.com/redis/redis/issues/135#issuecomment-2361996

We then delete any items from the sorted set with a "score" lower than
the current time. The Redis docs describe this a potentially slow
operation: https://redis.io/commands/zremrangebyscore/

If we don't want to call this on every request, we could instead add
some sort of random fuzzing to only occassionally call it. Or we could
abandon using a sorted set entirely and just store the idempotency key
as its own key in Redis using the standard Redis expiration
functionality. This would mean each request would store potentially
three keys (idempotency key, response key, status code key) instead of
the current two (response key, status code key).

Users who upgrade to a version of the library that includes this change
will see a failure in Redis when it tries to use a sortedset operation
on an existing set. At a minimum, this should be called out in a
changelog. Additionally, we could change the default idempotency key
name from `idempotency-key-keys` to something else to avoid a collision.

jmsanders commented:
I've added a new `expire_idempotency_keys` abstract method thatimplementations of Backend are responsible for implementing. The middleware calls it on every request.

For the Memory backend, we now store idempotency keys in a dict intead of a set. The value of each key is its expiration time.

We then iterate over all of the items in the dict and remove the ones with an expiration earlier than the current time. This operation isn't very efficient because it iterates over the entire dict, but given that the Memory backend is only meant to be used in local testing, I don't think it's worth optimizing further.

For the RedisBackend, we now store idempotency keys in a sortedset instead of a set. The "score" of each key is its expiration time. This matches the Redis project's official recommendation on how to expire items in a set: https://github.com/redis/redis/issues/135#issuecomment-2361996

We then delete any items from the sorted set with a "score" lower than the current time. The Redis docs describe this a potentially slow operation: https://redis.io/commands/zremrangebyscore/

If we don't want to call this on every request, we could instead add some sort of random fuzzing to only occassionally call it. Or we could abandon using a sorted set entirely and just store the idempotency key as its own key in Redis using the standard Redis expiration functionality. This would mean each request would store potentially three keys (idempotency key, response key, status code key) instead of the current two (response key, status code key).

Users who upgrade to a version of the library that includes this change will see a failure in Redis when it tries to use a sortedset operation on an existing set. At a minimum, this should be called out in a changelog. Additionally, we could change the default idempotency key name from `idempotency-key-keys` to something else to avoid a collision.

jmsanders commented on idempotency_header_middleware/backends/redis.py:
```diff
@@ -55,12 +56,20 @@ async def store_response_data(self, idempotency_key: str, payload: dict, status_
+        after the response expires.
+        """
+        if self.expiry:
+            await self.redis.zremrangebyscore(self.KEYS_KEY, '-inf', time.time())
```
I mentioned it in the main PR thread, but here's where it might make sense to instead store each idempotency key as its own key in Redis and avoid the potentially expensive [`ZREMRANGEBYSCORE`](https://redis.io/commands/zremrangebyscore/) call entirely.

sondrelg replied:
I'm not very concerned with this being a slow operation, as a slow redis operation is probably much faster for this type of use-case than making just one additional round-trip to the cache. At least that's what I'd expect.

If we wanted to store them separately and save the round-trips, I suppose we could use pipelining or a Lua script to perform the operations in one go. I'll leave that up to you 🙂

jmsanders committed:
Fix lint errors

sondrelg commented:
> Users who upgrade to a version of the library that includes this change will see a failure in Redis when it tries to use a sortedset operation on an existing set. At a minimum, this should be called out in a changelog. Additionally, we could change the default idempotency key name from idempotency-key-keys to something else to avoid a collision.

Is that because we re-use the same key value for the new data structure? Could we get around the error by changing the key? 

I guess upgrading will still effectively drop existing data, which isn't that great. Could this be mitigated if we stored multiple values?
